rust/dialog-storage/src/storage/backend/s3.rs

```diff
@@ -0,0 +1,2482 @@
+#[derive(Debug, Clone)]
+pub struct Get {
+    url: Url,
+}
+
+impl Get {
+    /// Create a new GET request for the given URL.
+    pub fn new(url: Url) -> Self {
+        Self { url }
+    }
+}
+
+impl Invocation for Get {
+    fn method(&self) -> &'static str {
+        "GET"
+    }
+
+    fn url(&self) -> &Url {
+        &self.url
+    }
+}
+impl Request for Get {}
+
+/// A DELETE request to remove an object.
+#[derive(Debug, Clone)]
+pub struct Delete {
+    url: Url,
+}
+
+impl Delete {
+    /// Create a new DELETE request for the given URL.
+    pub fn new(url: Url) -> Self {
+        Self { url }
+    }
+}
+
+impl Invocation for Delete {
+    fn method(&self) -> &'static str {
+        "DELETE"
+    }
+
+    fn url(&self) -> &Url {
+        &self.url
+    }
+}
+impl Request for Delete {}
+
+/// A GET request to list objects in a bucket.
+///
+/// Uses the S3 ListObjectsV2 API to retrieve object keys.
+#[derive(Debug, Clone)]
+pub struct List {
+    url: Url,
+}
+
+impl List {
+    /// Create a new list request for the given bucket URL with optional prefix.
+    ///
+    /// The URL should be the bucket root (e.g., `https://s3.amazonaws.com/bucket`)
+    /// with query parameters for `list-type=2` and optionally `prefix`.
+    pub fn new(mut url: Url, prefix: Option<&str>, continuation_token: Option<&str>) -> Self {
+        url.query_pairs_mut().append_pair("list-type", "2");
+        if let Some(prefix) = prefix {
+            url.query_pairs_mut().append_pair("prefix", prefix);
+        }
+        if let Some(token) = continuation_token {
+            url.query_pairs_mut()
+                .append_pair("continuation-token", token);
+        }
+        Self { url }
+    }
+}
+
+impl Invocation for List {
+    fn method(&self) -> &'static str {
+        "GET"
+    }
+
+    fn url(&self) -> &Url {
+        &self.url
+    }
+}
+impl Request for List {}
+
+/// Response from S3 ListObjectsV2 API (simplified).
+#[derive(Debug)]
+pub struct ListResult {
+    /// Object keys returned in this response.
+    pub keys: Vec<String>,
+    /// If true, there are more results to fetch.
+    pub is_truncated: bool,
+    /// Token to use for fetching the next page of results.
+    pub next_continuation_token: Option<String>,
+}
+
+/// S3-safe key encoding that preserves path structure.
+///
+/// Keys are treated as `/`-delimited paths. Each path component is checked:
+/// - If it contains only safe characters (alphanumeric, `-`, `_`, `.`), it's kept as-is
+/// - Otherwise, it's base58-encoded and prefixed with `!`
+///
+/// The `!` character is used as a prefix marker because it's in AWS S3's
+/// "safe for use" list and unlikely to appear at the start of path components.
+///
+/// Examples:
+/// - `remote/main` → `remote/main` (all components safe)
+/// - `remote/user@example` → `remote/!<base58>` (@ is unsafe, encode component)
+/// - `foo/bar/baz` → `foo/bar/baz` (all safe)
+pub fn encode_s3_key(bytes: &[u8]) -> String {
+    use base58::ToBase58;
```

Comment:
Nit: imports at the top of the module unless you have a mechanical reason to do otherwise.